### PR TITLE
docs: Update `getting-started.md`

### PR DESCRIPTION
**Pull request by JonathanWoollett-Light**
## Changes

Updates `getting-started.md`.

The example now involves setting up a microVM you can SSH into.

See https://github.com/JonathanWoollett-Light/firecracker/blob/getting-started/docs/getting-started.md.

## Reason

Some of the instructions are out-dated and it is missing a simple guide to ssh into a microVM.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by farazfazli** [comment]
Really appreciate the effort here in making the getting started guide easier to follow.

File: docs/getting-started.md [hunk 1]
```diff
@@ -1,372 +1,273 @@
...
-- **Linux**
-
-  Firecracker currently supports physical Linux **x86_64** and **aarch64**
-  hosts. The currently supported kernel versions can be found [here](kernel-policy.md).
+You can check if your system meets the requirements by running
+`firecracker/tools/devtool checkenv`.
 
-- **KVM**
+An opinionated way to run Firecracker is to launch an
+[EC2](https://aws.amazon.com/ec2/) `i3.metal` instance with Ubuntu 18.04.
```
Comment (roypat) [inline docs/getting-started.md:9]: nit: maybe we should recommend ubuntu 22, since that's what our build container uses? (also why is this "opinionated"? :D)

File: docs/getting-started.md [hunk 2]
```diff
@@ -1,372 +1,273 @@
...
 
-```bash
-git checkout tags/v0.10.1
-```
+# Download a rootfs
+wget https://s3.amazonaws.com/spec.ccfc.min/ci-artifacts/disks/${ARCH}/ubuntu-18.04.ext4
 
-### Building Firecracker
+# Download the ssh key for the rootfs
+wget https://s3.amazonaws.com/spec.ccfc.min/ci-artifacts/disks/${ARCH}/ubuntu-18.04.id_rsa
```
Comment (roypat) [inline docs/getting-started.md:76]: ah, didnt know this was something that needed to be done. explains all my difficulties ssh-ing into my microvms 😂

File: docs/getting-started.md [hunk 3]
```diff
@@ -1,372 +1,273 @@
...
+# Set network interface
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"iface_id\": \"net1\",
+        \"guest_mac\": \"$FC_MAC\",
+        \"host_dev_name\": \"$TAP_DEV\"
+    }" \
+    "http://localhost/network-interfaces/net1"
+
+sleep 0.015s
```
Comment (roypat) [inline docs/getting-started.md:191]: mh, instead of sleeping, can we give a bit more context here? Something along the lines of "Wait until you see in your first terminal that the microvm has started. Once it has done so, you can SSH into it using the key downloaded above, e.g [...]`?

File: docs/getting-started.md [hunk 4]
```diff
@@ -1,372 +1,273 @@
...
-to log into the guest machine. If you used our `hello-rootfs.ext4` image,
-you can login as `root`, using the password `root`.
-
-When you're done, issuing a `reboot` command inside the guest will actually
-shutdown Firecracker gracefully. This is due to the fact that Firecracker
-doesn't implement guest power management.
-
-**Note**: the default microVM will have 1 vCPU and 128 MiB RAM. If you wish to
-customize that (say, 2 vCPUs and 1024MiB RAM), you can do so before issuing
-the `InstanceStart` call, via this API command:
```
Comment (roypat) [inline docs/getting-started.md:210]: Are these defaults still mentioned anywhere? I think they're valuable information here

File: docs/getting-started.md (hunk 1 above)
Comment (roypat) [reply docs/getting-started.md:9]: Maybe briefly mention that we use `.metal` because we don't support nested virtualization?

**Review by roypat — COMMENTED** [review]
Looks good, as someone who's struggled through this guide multiple times these seem like improvements :D

File: docs/getting-started.md [hunk 5]
```diff
@@ -383,77 +284,3 @@ Points to check to confirm the API socket is running and accessible:
...
-or run Firecracker as `root` (via `sudo`).
-
-You can check if you have access to `/dev/kvm` with:
-
-```bash
-[ -r /dev/kvm ] && [ -w /dev/kvm ] && echo "OK" || echo "FAIL"
-```
-
-**Note:** If you've just added your user to the `kvm` group via `usermod`, don't
-forget to log out and then back in, so this change takes effect.
```
Comment (zulinx86) [inline docs/getting-started.md:415]: Why is this part removed?

File: docs/getting-started.md [hunk 6]
```diff
@@ -383,77 +284,3 @@ Points to check to confirm the API socket is running and accessible:
...
-
-## Appendix B: Setting Up Docker
```
Comment (zulinx86) [inline docs/getting-started.md:417]: We might need to discuss if we should remove this section.

**Review by zulinx86 — COMMENTED** [review]
I'd prefer to preserve both options "build from source" and "download latest release binaries".

File: docs/getting-started.md (hunk 5 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md:415]: This relates to standard Linux permissions, not specific to Firecracker. Although I don't have a strong view against keeping it.

File: docs/getting-started.md (hunk 1 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md:9]: Added a following comment:

> Firecracker does not support nested virtualization. This is why `.metal`
> instances are used.

File: docs/getting-started.md (hunk 3 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md:191]: Added comments on the `sleep`s:

> # API requests are handled asynchronously, it is important the configuration is
> # set, before `InstanceStart`.

and

> # API requests are handled asynchronously, it is important the microVM has been
> # started before we attempt to SSH into it.

respectively.

File: docs/getting-started.md (hunk 6 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md:417]: I don't think there is value in documenting the docker install process here.

**Review by roypat — APPROVED** [review]

File: docs/getting-started.md [hunk 7]
```diff
@@ -1,372 +1,282 @@
...
-curl -L ${release_url}/download/${latest}/firecracker-${latest}-${arch}.tgz \
-| tar -xz
-```
-
-Rename the binary to "firecracker":
-
-```console
-mv release-${latest}-$(uname -m)/firecracker-${latest}-$(uname -m) firecracker
+```bash
+lsmod | grep kvm
```
Comment (kalyazin) [inline docs/getting-started.md:26]: what if kvm is built in the kernel, not compiled as a module (I don't know how common is that)?

File: docs/getting-started.md [hunk 8]
```diff
@@ -1,372 +1,282 @@
...
+    "http://localhost/logger"
+
+# Set machine config
+#
+# Without this the microVM will have the default values of 1 vCPU and 128MB of RAM.
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"vcpu_count\": 2,
+        \"mem_size_mib\": 128,
+        \"cpu_template\": \"C3\"
```
Comment (kalyazin) [inline docs/getting-started.md:157]: do we want our users to jump straight into a template?

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:157]: Good point, do you think this should just be removed?

Comment (kalyazin) [reply docs/getting-started.md:157]: I think it's an advanced feature, I wouldn't encourage using it by default.

Comment (JonathanWoollett-Light) [reply docs/getting-started.md]: Okay, I've removed this API call.

File: docs/getting-started.md (hunk 7 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md:26]: I'm not aware of this case, do you have any documentation of this you can link?

Comment (kalyazin) [reply docs/getting-started.md:26]: I think, checking for the device file is ok (already mentioned below).
Other ways are hacky (like grepping `/proc/kallsyms`).
Maybe leave the `lsmod` check as is, but also add a note that kvm may be compiled in and still be used, but not visible via `lsmod`?

Comment (kalyazin) [reply docs/getting-started.md:26]: Looks good to me.
```
config KVM
        tristate "Kernel-based Virtual Machine (KVM) support"
```
Tristate usually means enabled, disabled or module.

File: docs/getting-started.md (hunk 1 above)
Comment (zulinx86) [reply docs/getting-started.md:9]: To my understanding, the reason why we use `.metal` instances is that firecracker requires KVM access and non-metal instances do not support nested virtualization.

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:9]: Updated comment to:

> Nested virtualization is not supported on non-metal EC2 instances. This is why
> `.metal` instances are used.

File: docs/getting-started.md [hunk 9]
```diff
@@ -1,372 +1,271 @@
...
+sudo sh -c "echo 1 > /proc/sys/net/ipv4/ip_forward"
+sudo iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE || true
+sudo iptables -D FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT \
+    || true
+sudo iptables -D FORWARD -i tap0 -o eth0 -j ACCEPT || true
+
+# Enable microVM internert access
+sudo iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
+sudo iptables -I FORWARD 1 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+sudo iptables -I FORWARD 1 -i tap0 -o eth0 -j ACCEPT
```
Comment (zulinx86) [inline docs/getting-started.md:132]: The first three `iptables` command for deleting existing rules.
```suggestion
# Enable ip forwarding
sudo sh -c "echo 1 > /proc/sys/net/ipv4/ip_forward"

# Set up microVM internet access
sudo iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE || true
sudo iptables -D FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT \
    || true
sudo iptables -D FORWARD -i tap0 -o eth0 -j ACCEPT || true
sudo iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
sudo iptables -I FORWARD 1 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
sudo iptables -I FORWARD 1 -i tap0 -o eth0 -j ACCEPT
```

File: docs/getting-started.md [hunk 10]
```diff
@@ -1,372 +1,271 @@
...
+KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd
+ 8250.nr_uarts=1 random.trust_cpu=on"
+
+# Set boot source
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"kernel_image_path\": \"${KERNEL}\",
+        \"boot_args\": \"${KERNEL_BOOT_ARGS}\"
+    }" \
+    "http://localhost/boot-source"
```
Comment (zulinx86) [inline docs/getting-started.md:160]: Trying on AL2, I got this error.
```
$ curl -X PUT --unix-socket "${API_SOCKET}" \
>     --data "{
>         \"kernel_image_path\": \"${KERNEL}\",
>         \"boot_args\": \"${KERNEL_BOOT_ARGS}\"
>     }" \
>     "http://localhost/boot-source"
{"fault_message":"An error occurred when deserializing the json body of a request: control character (\\u0000-\\u001F) found while parsing a string at line 4 column 0."}
```

File: docs/getting-started.md (hunk 9 above)
Comment (JonathanWoollett-Light) [reply docs/getting-started.md]: Fixed.

File: docs/getting-started.md (hunk 10 above)
Comment (zulinx86) [reply docs/getting-started.md:160]: I could fix this by changing:

before
```
KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd
 8250.nr_uarts=1 random.trust_cpu=on"
```

after
```
KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd 8250.nr_uarts=1 random.trust_cpu=on"
```

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:160]: Does the previous API request of :
```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"log_path\": \"${LOGFILE}\",
        \"level\": \"Debug\",
        \"show_level\": true,
        \"show_log_origin\": true
    }" \
    "http://localhost/logger"
```
succeed?

What does `echo $KERNEL` and `echo $KERNEL_BOOT_ARGS` give?

File: docs/getting-started.md [hunk 11]
```diff
@@ -1,372 +1,271 @@
...
+
+# API requests are handled asynchronously, it is important the microVM has been
+# started before we attempt to SSH into it.
+sleep 0.015s
+
+# SSH into the microVM
+sudo ssh -i ./ubuntu-18.04.id_rsa 172.16.0.2
+
+# Use `root` for both the login and password.
+# Run `reboot` to exit.
```
Comment (zulinx86) [inline docs/getting-started.md:204]: Although this includes instructions for net configuration on host, but does not include instructions for net configuration inside guests. How about adding it or redirect to https://github.com/firecracker-microvm/firecracker/blob/main/docs/network-setup.md ?

File: docs/getting-started.md (hunk 10 above)
Comment (zulinx86) [reply docs/getting-started.md:160]: > Does the previous API request of:
> 
> ```shell
> curl -X PUT --unix-socket "${API_SOCKET}" \
>     --data "{
>         \"log_path\": \"${LOGFILE}\",
>         \"level\": \"Debug\",
>         \"show_level\": true,
>         \"show_log_origin\": true
>     }" \
>     "http://localhost/logger"
> ```
> 
> succeed?
> 
> What does `echo $KERNEL` and `echo $KERNEL_BOOT_ARGS` give?


Yes.
Here is the log.
```
[ec2-user@ip-172-31-11-151 ~]$ curl -X PUT --unix-socket "${API_SOCKET}" \
>     --data "{
>         \"log_path\": \"${LOGFILE}\",
>         \"level\": \"Debug\",
>         \"show_level\": true,
>         \"show_log_origin\": true
>     }" \
>     "http://localhost/logger"
[ec2-user@ip-172-31-11-151 ~]$ KERNEL="./vmlinux.bin"
[ec2-user@ip-172-31-11-151 ~]$ KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd
>  8250.nr_uarts=1 random.trust_cpu=on"
[ec2-user@ip-172-31-11-151 ~]$ curl -X PUT --unix-socket "${API_SOCKET}" \
>     --data "{
>         \"kernel_image_path\": \"${KERNEL}\",
>         \"boot_args\": \"${KERNEL_BOOT_ARGS}\"
>     }" \
>     "http://localhost/boot-source"
{"fault_message":"An error occurred when deserializing the json body of a request: control character (\\u0000-\\u001F) found while parsing a string at line 4 column 0."}
[ec2-user@ip-172-31-11-151 ~]$ echo $KERNEL
./vmlinux.bin
[ec2-user@ip-172-31-11-151 ~]$ echo $KERNEL_BOOT_ARGS
console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd 8250.nr_uarts=1 random.trust_cpu=on
```

Comment (zulinx86) [reply docs/getting-started.md:160]: I'm suspecting that `KERNEL_BOOT_ARGS` includes invisible control character...

Comment (zulinx86) [reply docs/getting-started.md:160]: Or something happens when copying and pasting it. Any way, removing the new line works on my env.
```
[ec2-user@ip-172-31-11-151 ~]$ KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 ipv6.disable=1 i8042.nokbd 8250.nr_uarts=1 random.trust_cpu=on"
[ec2-user@ip-172-31-11-151 ~]$ curl -X PUT --unix-socket "${API_SOCKET}"     --data "{
        \"kernel_image_path\": \"${KERNEL}\",
        \"boot_args\": \"${KERNEL_BOOT_ARGS}\"
    }"     "http://localhost/boot-source"
[ec2-user@ip-172-31-11-151 ~]$
```

File: docs/getting-started.md (hunk 11 above)
Comment (zulinx86) [reply docs/getting-started.md:204]: Ah sorry, this seems to be caused by the conflict of ip address range for host and guest.
```
$ ip route
default via 172.31.0.1 dev eth0
169.254.169.254 dev eth0
172.16.0.0/30 dev tap0 proto kernel scope link src 172.16.0.1
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown
172.31.0.0/20 dev eth0 proto kernel scope link src 172.31.11.151
```

File: docs/getting-started.md [hunk 12]
```diff
@@ -1,372 +1,270 @@
...
+sudo ip addr add "${TAP_IP}${MASK_SHORT}" dev "$TAP_DEV"
+sudo ip link set dev "$TAP_DEV" up
+
+# Enable ip forwarding
+sudo sh -c "echo 1 > /proc/sys/net/ipv4/ip_forward"
+
+# Set up microVM internet access
+sudo iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE || true
+sudo iptables -D FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT \
+    || true
```
Comment (zulinx86) [inline docs/getting-started.md:128]: nit: remove the new line as well?

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:128]: The newline is needed to conform to [`mdl`](https://github.com/markdownlint/markdownlint) column max width.

File: docs/getting-started.md (hunk 11 above)
Comment (zulinx86) [reply docs/getting-started.md:206]: The guest decides its ip address based on MAC address with [fcnet-setup.sh](https://github.com/firecracker-microvm/firecracker/blob/main/resources/tests/fcnet-setup.sh).

Like my case, in some customers' environment, address range conflicts can happen.
To make `TAP_IP` modification acceptable, it would be nice that `FC_MAC` is derived from `TAP_IP` by doing reverse calculation of `fcnet-setup.sh`.

Comment (zulinx86) [reply docs/getting-started.md:206]: Here is an example.
```
FC_MAC=$(printf "06:00:%02x:%02x:%02x:%02x" $(echo $TAP_IP | sed "s/\./ /g"))
```

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:206]: Would replacing `FC_MAC` with:
```bash
# Derive FC_MAC from TAP_IP e.g. `172.16.0.1` becomes `06:00:AC:10:00:02`.
FC_MAC=$(printf "06:00:%02x:%02x:%02x:%02x" $(echo $TAP_IP | sed "s/\./ /g"))
```
fit this?

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:209]: Included comment on this:
```bash
# The IP address of a guest is derived from its MAC address with
# `fcnet-setup.sh`, this has been pre-configured in the guest rootfs. It is
# important that `TAP_IP` and `FC_MAC` match this.
FC_MAC="06:00:AC:10:00:02"
```

File: docs/getting-started.md (hunk 10 above)
Comment (zulinx86) [reply docs/getting-started.md:162]: talked with Jonathan and concluded to simpify it.

File: docs/getting-started.md [hunk 13]
```diff
@@ -1,372 +1,276 @@
...
+        \"show_level\": true,
+        \"show_log_origin\": true
+    }" \
+    "http://localhost/logger"
+
+KERNEL="./vmlinux.bin"
+KERNEL_BOOT_ARGS="console=ttyS0 reboot=k panic=1 pci=off"
+
+if [ ${ARCH} = "aarch64" ]; then
+    KERNEL_BOOT_ARGS="keep_bootcon ${KERNEL_BOOT_ARGS}"
```
Comment (zulinx86) [inline docs/getting-started.md:154]: In this terminal, `$ARCH` is not set.

Comment (zulinx86) [reply docs/getting-started.md:154]:
```suggestion
ARCH=$(uname -m)
if [ ${ARCH} == "aarch64" ]; then
    KERNEL_BOOT_ARGS="keep_bootcon ${KERNEL_BOOT_ARGS}"
fi
```

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:154]: Fixed.

File: docs/getting-started.md [hunk 14]
```diff
@@ -1,372 +1,279 @@
...
+
+# API requests are handled asynchronously, it is important the microVM has been
+# started before we attempt to SSH into it.
+sleep 0.015s
+
+# SSH into the microVM
+sudo ssh -i ./ubuntu-18.04.id_rsa 172.16.0.2
+
+# Use `root` for both the login and password.
+# Run `reboot` to exit.
```
Comment (zulinx86) [inline docs/getting-started.md:212]: To get network access from guest, it requires to set up default gateway.
https://github.com/firecracker-microvm/firecracker/blob/main/docs/network-setup.md?plain=1#L89

Comment (zulinx86) [reply docs/getting-started.md:212]: And it would be better to have DNS setup as well.

File: docs/getting-started.md (hunk 13 above)
Comment (zulinx86) [reply docs/getting-started.md:156]: I think it needs `==`.

Comment (JonathanWoollett-Light) [reply docs/getting-started.md:156]: It is `if [ ${arch} = "x86_64" ]; then` in the previous getting started documentation, and in bash comparison is typically `=` https://tldp.org/LDP/abs/html/comparison-ops.html. Could you maybe explain more why it would be `==` here?

Comment (zulinx86) [reply docs/getting-started.md:156]: You're right. My bad.

**Review by zulinx86 — APPROVED** [review]

File: docs/getting-started.md (hunk 14 above)
Comment (zulinx86) [reply docs/getting-started.md:212]: Talking with Jonathan offline, the connectivity between guests and the internet can be considered as optional and can be handled in a follow-up PR. Resolving this conversation.

**Review by kalyazin — APPROVED** [review]
